src/background/redirect-protection.js

```diff
@@ -0,0 +1,302 @@
+        !details.url.includes('redirect-protection')
+      ) {
+        chrome.storage.session
+          .set({
+            [getRedirectUrlStorageKey(details.tabId)]: details.url,
+          })
+          .catch((err) => {
+            console.error('[redirect-protection] Failed to store URL:', err);
+          });
+      }
+    },
+    {
+      url: [{ schemes: ['http', 'https'] }],
+    },
+  );
+
+  chrome.tabs.onRemoved.addListener((tabId) => {
+    chrome.storage.session
+      .remove(getRedirectUrlStorageKey(tabId))
+      .catch(() => {});
+
+    chrome.declarativeNetRequest
+      .updateSessionRules({
+        removeRuleIds: [REDIRECT_PROTECTION_SESSION_ID_RANGE.start + tabId],
+      })
+      .catch(() => {});
+  });
+
+  chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
+    if (message.action === 'getRedirectUrl') {
+      if (sender.tab && sender.tab.id) {
+        const storageKey = getRedirectUrlStorageKey(sender.tab.id);
+        chrome.storage.session
+          .get(storageKey)
+          .then((result) => {
+            sendResponse({
+              url: result[storageKey] || null,
+            });
+          })
+          .catch(() => {
+            sendResponse({ url: null });
+          });
+        return true;
+      }
+      sendResponse({ url: null });
+      return false;
+    }
+
+    if (message.action === 'allowRedirect') {
+      if (!sender.tab || !sender.tab.id || !message.url) {
+        console.error(
+          '[redirect-protection] Missing tab or URL in allowRedirect',
+        );
+        sendResponse({ success: false, error: 'Missing tab or URL' });
+        return false;
+      }
+
+      const tabId = sender.tab.id;
+      const url = message.url;
+
+      (async () => {
+        try {
+          const urlObj = new URL(url);
+          const urlPattern = `||${urlObj.host}${urlObj.pathname}${urlObj.search}${urlObj.hash}`;
+
+          await chrome.declarativeNetRequest.updateSessionRules({
+            addRules: [
+              {
+                id: REDIRECT_PROTECTION_SESSION_ID_RANGE.start + tabId,
+                priority: REDIRECT_PROTECTION_SESSION_PRIORITY,
+                action: { type: 'allow' },
+                condition: {
+                  urlFilter: urlPattern,
+                  resourceTypes: ['main_frame'],
+                  tabIds: [tabId],
+                },
+              },
+            ],
+            removeRuleIds: [REDIRECT_PROTECTION_SESSION_ID_RANGE.start + tabId],
+          });
+
+          await chrome.storage.session.remove(getRedirectUrlStorageKey(tabId));
+
+          sendResponse({ success: true });
+        } catch (error) {
+          console.error(
+            '[redirect-protection] Error creating session rule:',
+            error,
+          );
+          sendResponse({ success: false, error: error.message });
+        }
+      })();
+
+      return true;
+    }
+
+    if (message.action === 'disableRedirectProtection') {
```

Comment:
This should be done in place in `pages/redirect-protection` as they have access to `Options`.

The whole update of DNR, etc.. should be reaction to the newest state of the `Options`.